### PR TITLE
CODAP-1286: relational operators propagate empty values

### PR DESCRIPTION
**Pull request by kswenson**
## Summary
- Relational operators (`<`, `<=`, `>`, `>=`) propagate empty as `""` and NaN as NaN, matching V2 (apps/dg/formula/formula.js). Four near-duplicate operator bodies refactored into a shared `compareValues` helper.
- Render layer (`renderAttributeValue` + `getNumFormatter`) hides NaN for numeric attributes only. Categorical/text attributes preserve user-typed `"NaN"`; Infinity continues to render as `"Infinity"` (still potentially meaningful, unlike NaN).

Fixes CODAP-1286

## Test plan
- [ ] Formula `x < 30` returns empty for cases where x is empty (case table cell renders blank)
- [ ] Formula `1/0` still renders as `Infinity`
- [ ] Formula `0/0` renders as empty in a numeric attribute
- [ ] Literal `"NaN"` typed into a categorical/text attribute is preserved
- [ ] Cypress regression suite passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: v3/src/models/formula/functions/operators.ts [hunk 1]
```diff
@@ -22,6 +22,30 @@ function modError(a: any, b: any) {
...
 
+// Shared logic for the relational operators (<, <=, >, >=). Matches V2 (apps/dg/formula/formula.js)
+// in propagating empty values as "" and NaN values as NaN. Empty propagates against numeric or
+// empty operands; an empty paired with a non-numeric string falls through to lexical comparison.
+function compareValues(a: any, b: any, compare: (x: number | string, y: number | string) => boolean) {
+  if (Number.isNaN(a) || Number.isNaN(b)) return NaN
+  const isAEmpty = isValueEmpty(a)
+  const isBEmpty = isValueEmpty(b)
+  if (isAEmpty && isBEmpty) return ""
+  // Fast path: both arguments are already numbers — skip expensive date parsing
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/functions/operators.ts:33]: Good catch. The downstream issue was real: `if(NaN, ...)` was hitting `Number(NaN) !== 0 → true` and taking the truthy branch, and `isValueTruthy(NaN)` was returning `true` for aggregate filters. Both diverged from V2 (which uses `!!NaN === false` semantics).

Fixed in a follow-up commit:
- `isValueTruthy` now treats NaN as falsy (uses strict `Number.isNaN` so non-numeric strings like `"abc"` remain truthy).
- `if()` short-circuits to `falseValue` for NaN conditions before the legacy `Number(condition) !== 0` check.

Added regression tests in `function-utils.test.ts` (`isValueTruthy(NaN) === false`) and `other-functions.test.ts` (`if(0/0 < 0, ...)` returns the false branch).

File: v3/src/models/formula/functions/operators.ts [hunk 2]
```diff
@@ -22,6 +22,30 @@ function modError(a: any, b: any) {
...
+  if (isADate) a = aDate.valueOf() / 1000
+  if (isBDate) b = bDate.valueOf() / 1000
+  // compare numerically if possible
+  const [isANumber, aNumber] = checkNumber(a)
+  const [isBNumber, bNumber] = checkNumber(b)
+  if (isANumber && isBNumber) return compare(aNumber, bNumber)
+  // empty paired with a numeric value → propagate empty (V2 parity)
+  if ((isAEmpty && isBNumber) || (isANumber && isBEmpty)) return ""
+  // compare as strings
+  return compare(String(a), String(b))
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/functions/operators.ts]: Good catch — accepted. The fall-through now normalizes empty operands to "" before the lexical comparison so `null < 'abc'` and `undefined < 'abc'` behave the same as `'' < 'abc'` (true) instead of stringifying as `"null"`/`"undefined"`. Added regression tests in operators.test.ts for both null and undefined operand cases.

Worth noting: V2's `stringFinisher` (apps/dg/formula/formula.js:344) falls through to the same `String(...)` conversion and has the same edge case — this is a small improvement over strict V2 parity.

File: v3/src/models/formula/functions/operators.test.ts [hunk 3]
```diff
@@ -65,11 +76,18 @@ describe("<= operator", () => {
     expect(f2.evaluate()).toBe(true)
     const f3 = math.compile("2 <= 1")
     expect(f3.evaluate()).toBe(false)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/functions/operators.test.ts:78]: Good catch — accepted. Pre-existing typo (used `<` instead of `<=` in the third string-comparison assertion); fixed in this commit so the assertion actually exercises `<=`.

File: v3/src/models/formula/functions/function-utils.ts [hunk 4]
```diff
@@ -9,7 +9,9 @@ export const UNDEF_RESULT = ""
 
 // CODAP formulas assume that 0 is a truthy value, which is different from default JS behavior. So that, for instance,
 // count(attribute) will return a count of valid data values, since 0 is a valid numeric value.
-export const isValueTruthy = (value: any) => isValueNonEmpty(value) && value !== false
+// NaN is treated as falsy (V2 parity: V2 uses !!x where !!NaN === false) so a filter expression
+// evaluating to NaN excludes that case from aggregations.
+export const isValueTruthy = (value: any) => isValueNonEmpty(value) && value !== false && !Number.isNaN(value)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/functions/function-utils.ts:15]: Good catch — accepted. Applied the same fix pattern as `if()` and `isValueTruthy`: strict `Number.isNaN(arg)` check before the legacy `Number(arg) !== 0` fall-through. `boolean(0/0)` now returns false. Updated the existing `boolean(NaN)` test that asserted true.

File: v3/src/components/case-tile-common/attribute-format-utils.tsx [hunk 5]
```diff
@@ -130,11 +133,19 @@ export function renderAttributeValue(str = "", num = NaN, attr?: IAttribute, opt
...
   }
 
-  // numbers
-  if (isFinite(num)) {
+  // numbers — finite values for any attribute, plus NaN for numeric and formula attributes
+  // (where NaN comes from computation rather than user-typed text and should render as empty).
+  // Formula attributes are included because their type may be unset or non-numeric when every
+  // computed value is NaN (e.g., a "0/0" formula). Infinity falls through to the default text
+  // branch so the literal "Infinity" remains visible — still potentially meaningful, unlike NaN.
+  const isComputedNaN = attr?.hasFormula && str === "NaN" && Number.isNaN(num)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/case-tile-common/attribute-format-utils.tsx]: Acknowledged but not addressed in this PR. This is a fundamental V3 architectural limitation: V3 stores all case values as strings, so a formula returning JS `NaN` and a formula returning the literal string `"NaN"` end up indistinguishable at the render layer (`importValueToString(NaN)` → `"NaN"`, `toNumeric("NaN")` → `NaN`, identical to `importValueToString("NaN")`). V2 doesn't have this problem because `cellValue` retains its JS type end-to-end.

Tightening to `type === "numeric"` would regress the `0/0` case this PR was meant to fix (an all-NaN formula has `type === undefined`). Tightening to `userType === "numeric"` would force users to opt-in explicitly. Given that returning the literal string `"NaN"` from a formula is contrived (vs. `0/0` or empty-comparison NaN being common), we're accepting the edge case as a known limitation.

**Review by tealefristoe — APPROVED** [review]
Looks good 👍
